New Resource: `azurerm_container_registry_credential_set`

jan-mrm opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Adding a new resource in order to manage Credential Sets for Azure Container Registries.


## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

none

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

```
make acctests SERVICE='containers' TESTARGS='-run=TestAccContainerRegistryCredentialSet_' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/containers -run=TestAccContainerRegistryCredentialSet_ -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccContainerRegistryCredentialSet_basic
=== PAUSE TestAccContainerRegistryCredentialSet_basic
=== RUN   TestAccContainerRegistryCredentialSet_requiresImport
=== PAUSE TestAccContainerRegistryCredentialSet_requiresImport
=== RUN   TestAccContainerRegistryCredentialSet_update
=== PAUSE TestAccContainerRegistryCredentialSet_update
=== CONT  TestAccContainerRegistryCredentialSet_basic
=== CONT  TestAccContainerRegistryCredentialSet_update
=== CONT  TestAccContainerRegistryCredentialSet_requiresImport
--- PASS: TestAccContainerRegistryCredentialSet_basic (120.32s)
--- PASS: TestAccContainerRegistryCredentialSet_requiresImport (122.08s)
--- PASS: TestAccContainerRegistryCredentialSet_update (156.95s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/containers    159.627s
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_container_registry_credential_set` - New resource for managing Credential Sets of Container Registries [GH-27528]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Closes #26539


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

stephybun commented on website/docs/r/container_registry_credential_set.html.markdown:
```diff
@@ -0,0 +1,148 @@
+
+A `auth_credentials` block supports the following:
+
+* `username_secret_identifier` - (Required) The URI of the secret containing the user's name in a Key Vault.
+
+* `password_secret_identifier` - (Required) The URI of the secret containing the user's password in a Key Vault.
```
```suggestion
* `username_secret_identifier` - (Required) The URI of the secret containing the username in a Key Vault.

* `password_secret_identifier` - (Required) The URI of the secret containing the password in a Key Vault.
```

jan-mrm commented on internal/services/containers/container_registry_credential_set_resource.go:
```diff
@@ -0,0 +1,331 @@
+					LoginServer:     pointer.To(config.LoginServer),
+					AuthCredentials: expandAuthCredentials(config.AuthCredentials),
+				},
+				Identity: &identity.SystemAndUserAssignedMap{
+					Type: identity.TypeSystemAssigned,
+				},
```
Hey, the [spec](https://github.com/hashicorp/terraform-provider-azurerm/blob/e2798fc44e42e245ce3ade8b1a6511be85c710dd/vendor/github.com/hashicorp/go-azure-sdk/resource-manager/containerregistry/2023-07-01/credentialsets/model_credentialset.go#L13C1-L14C1) seems to allow a System- or UserAssigned Identity but when I tried I gotten this error:

```
{"error": {
    "code": "CannotSetResourceIdentity",
    "message": "Resource type 'Microsoft.ContainerRegistry/registries/credentialSets' does not support creation of 'UserAssigned' resource identity. The supported types are 'SystemAssigned'."
}}
```
What would you propose?
1. Expose the `identity` to the user and note in the docs that only SystemAssigned is an allowed value (also default to this value of cause) - I guess the idea would be to have it extendable already if the user someday might be allowed to set a UserAssigned Identity
2. keep it as is? So set the identity hardcoded to SystemAssgined and only return it as a read-only property
3. other ideas?
 
Currently I'm still getting errors locally after I swapped from my defined Identity to the `SystemAndUserAssignedMap`. I'm still investigating.

jan-mrm replied:
@stephybun thanks for your review, I just pushed an updated version with all your feedback implemented. This point remains open for discussion. Currently I updated the object to the returned identity.SystemOrUserAssignedIdentity but still as read-only. Awaiting your feedback 🙂

stephybun replied:
@jan-mrm, since it's taken me quite a while to circle back to this (sorry! 🙈) would you be able to confirm whether `UserAssigned` identity is still unsupported for this resource?

jan-mrm replied:
@stephybun thanks for coming back 🙂, I'll check it again

jan-mrm replied:
hey again @stephybun , just tried it using the api (2023-01-01-preview) and I get the following errors:

```json
{
  "error": {
    "code": "CannotSetResourceIdentity",
    "message": "Resource type 'Microsoft.ContainerRegistry/registries/credentialSets' does not support creation of 'UserAssigned' resource identity. The supported types are 'SystemAssigned'."
  }
}
```

```json
{
  "error": {
    "code": "CannotSetResourceIdentity",
    "message": "Resource type 'Microsoft.ContainerRegistry/registries/credentialSets' does not support creation of 'SystemAssigned, UserAssigned' resource identity. The supported types are 'SystemAssigned'."
  }
}
```
And actually the following when providing an empty itentity.type string:
```json
{
  "error": {
    "code": "OnlySystemManagedIdentityAllowed",
    "message": "Only System Managed Identities are allowed for resources of type 'Microsoft.ContainerRegistry/registries/credentialSets'. For more information, please visit https://aka.ms/acr/cache."
  },
  "status": "Failed"
}
```
So we could have it as read-only like its currently or "expose" the field but only allow the value `SystemAssigned` for now.

jan-mrm replied:
@stephybun done that. Feel free to review once again.

I actually didn't need the custom expand/flatten functions as far as I can see now.

Last thing now is: can one overwrite the validation function and only allow SystemAssigned although using `SystemAssignedIdentityRequired`? Or do you think thats bad or doesn't even work in the first place? I feel like catching that when running 'plan' is nicer than when running 'apply' 🙂 (anyways I did add a note to the implementation and the docs).

jan-mrm replied:
@stephybun Hey, sorry I total misread your comment. Now I changed it to `SystemAssignedIdentityRequired` and added flatten and expand functions to convert the types.
Sadly I do face one more issue now... a diff after an apply:
```
      ~ identity {
          ~ type         = "systemAssigned" -> "SystemAssigned"
            # (2 unchanged attributes hidden)
        }
```
The API returns a lower case "systemAssigned" which is passed through even though its of type `identity.Type`, I'm not really sure why. Got a solution for me how to fix that? 😅
Guess I could hardcode the identity type in the expand and flatten functions to identity.TypeSystemAssigned?
(I don't think I had that issue when i used SystemAssignedUserAssignedIdentityRequired, I'd guess the functions like identity.FlattenSystemAndUserAssignedMapToModel have a case-conversion - I'll take a look into that in a bit)

stephybun replied:
@jan-mrm see my latest review for my suggestion on this

jan-mrm commented on internal/services/containers/container_registry_credential_set_resource.go:
```diff
@@ -0,0 +1,331 @@
+					},
+				},
+			},
+		},
```
updated to `"identity": commonschema.SystemOrUserAssignedIdentityComputed(),` since the spec returns that type

stephybun commented on internal/services/containers/container_registry_credential_set_resource.go:
```diff
@@ -0,0 +1,304 @@
+	}
+	output := make([]identity.ModelSystemAssigned, 1)
+	output[0] = identity.ModelSystemAssigned{
+		Type:        input.Type,
```
We call a normalising function on the identity type returned here in the flatten functions of the `identity` package because the API is often inconsistent with the casing

https://github.com/hashicorp/go-azure-helpers/blob/624ef4f92d9f94a4a467b2bf5fb79974ef6c3255/resourcemanager/identity/constants.go#L20-L38

We can do this and it should take care of the casing issue for us 
```suggestion
		Type:        identity.Type(input.Type),
```

jan-mrm replied:
I tested that and it did not resolve the issue 🫤

stephybun commented on internal/services/containers/container_registry_credential_set_resource_test.go:
```diff
@@ -0,0 +1,173 @@
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		{
+			Config:      r.requiresImport(data),
+			ExpectError: acceptance.RequiresImportError("azurerm_container_registry_credential_set"),
+		},
```
We can replace this with 
```suggestion
		data.RequiresImportErrorStep(r.requiresImport),
```

stephybun commented on internal/services/containers/container_registry_credential_set_resource_test.go:
```diff
@@ -0,0 +1,173 @@
+    type = "SystemAssigned"
+  }
+  authentication_credentials {
+    username_secret_id = "https://example-keyvault.vault.azure.net/secrets/acr-cs-user-name"
+    password_secret_id = "https://example-keyvault.vault.azure.net/secrets/acr-cs-user-password"
```
Running the test locally fails because this doesn't exist, we can you add a test template that creates the dependencies for this acceptance test?

stephybun commented:
@jan-mrm I've opened a PR on Pandora which will patch the SDK to match the behaviour we see in the API (instead of what's been defined in the Rest API spec). Once that's made it through to [hashicorp/go-azure-sdk](https://github.com/hashicorp/go-azure-sdk) we can bump the version in the provider, rebase this PR on top of main and get rid of the custom expand/flatten which will resolve the casing issue.

jan-mrm commented:
@stephybun thanks for the workaround implementation. I updated the go-azure-sdk and updated this branch. Local testing looks good on my end and the new acctests pass locally.

stephybun commented on internal/services/containers/container_registry_credential_set_resource.go:
```diff
@@ -0,0 +1,286 @@
+				},
+			},
+		},
+		// Note [1]: At point in time of the implementation of this resource the API only accept SystemAssigned even though API Spec defines all three identity modes are possible
+		// Error when trying with type UserAssigned:
+		//	code: "CannotSetResourceIdentity"
+		//	message: "The resource identity 'UserAssigned' cannot be set on the resource of type 'Microsoft.ContainerRegistry/registries/credentialSets'."
+		// or with type empty:
+		//	code: "OnlySystemManagedIdentityAllowed"
+		//	message: "Only System Managed Identities are allowed for resources of type 'Microsoft.ContainerRegistry/registries/credentialSets'. For more information, please visit https://aka.ms/acr/cache."
```
Can we update this comment to link to the swagger issue?

```suggestion
		// This property relies on a pandora workaround due to a swagger issue
		// https://github.com/Azure/azure-rest-api-specs/issues/32154
```

stephybun reviewed (approved):
Thanks for your effort and patience here @jan-mrm LGTM 🚀
